src/prng/isaac.rs

```diff
@@ -333,41 +331,33 @@ fn mix(a: &mut w32, b: &mut w32, c: &mut w32, d: &mut w32,
     *h ^= *a >> 9;  *c += *h; *a += *b;
 }
 
-impl Rand for IsaacRng {
-    fn rand<R: Rng>(other: &mut R) -> IsaacRng {
-        let mut key = [w(0); RAND_SIZE];
-        unsafe {
-            let ptr = key.as_mut_ptr() as *mut u8;
+impl SeedableRng for IsaacRng {
+    type Seed = [u8; 32];
 
-            let slice = slice::from_raw_parts_mut(ptr, RAND_SIZE * 4);
-            other.fill_bytes(slice);
+    fn from_seed(seed: Self::Seed) -> Self {
+        let mut seed_u32 = [0u32; 8];
+        le::read_u32_into(&seed, &mut seed_u32);
+        let mut seed_extended = [w(0); RAND_SIZE];
+        for (x, y) in seed_extended.iter_mut().zip(seed_u32.iter()) {
+            *x = w(*y);
         }
-
-        init(key, 2)
-    }
-}
-
-impl<'a> SeedableRng<&'a [u32]> for IsaacRng {
-    fn reseed(&mut self, seed: &'a [u32]) {
-        *self = Self::from_seed(seed);
+        init(seed_extended, 2)
     }
-    /// Create an ISAAC random number generator with a seed. This can
-    /// be any length, although the maximum number of elements used is
-    /// 256 and any more will be silently ignored. A generator
-    /// constructed with a given seed will generate the same sequence
-    /// of values as all other generators constructed with that seed.
-    fn from_seed(seed: &'a [u32]) -> IsaacRng {
-        let mut key = [w(0); RAND_SIZE];
 
-        // make the seed into [seed[0], seed[1], ..., seed[seed.len()
-        // - 1], 0, 0, ...], to fill `key`.
-        let seed_iter = seed.iter().map(|&x| x).chain(repeat(0u32));
+    fn from_rng<R: Rng>(mut rng: R) -> Result<Self, Error> {
+        // Custom `from_rng` implementations that fills the entire state
+        let mut seed = [w(0u32); RAND_SIZE];
+        unsafe {
+            let ptr = seed.as_mut_ptr() as *mut u8;
 
-        for (rsl_elem, seed_elem) in key.iter_mut().zip(seed_iter) {
-            *rsl_elem = w(seed_elem);
+            let slice = slice::from_raw_parts_mut(ptr, RAND_SIZE * 4);
+            rng.try_fill_bytes(slice)?;
+        }
+        for i in seed.iter_mut() {
+            *i = w(i.0.to_le());
```

Comment:
Yes, you argued for it 😄. I don't mind, it will only run on Big-Endian and should not make it much slower.

The array is 2 KiB for ISAAC-64, maybe we don't want to increase it to 4 KiB on the stack by using two arrays? Probably not going to be a problem though.

Reply:
You're right on both accounts, though maybe I was wrong to argue for the Endian-fixing, but it doesn't matter much.

But, since this sets the whole of `mem`, why not call `init(seed, 0)`?

Reply:
We could. I still feel a bit uneasy about it though. We are tying the security of one RNG (ISAAC in this case) to the security of the seeding RNG. Using the seeding function provided with the RNG will keep it 'pure', any security promises for ISAAC are not influenced by the choice of seeding RNG (provided it is statistically good).